Basic/DyanmicPrefabs/Assets/Scripts/04_Server Authoritative Spawn As Network-Invisible For Loading Clients/SpawnPrefabInvisibleWhileIsLoading.cs

```diff
@@ -0,0 +1,154 @@
+        
+        // invoked by UI
+        public void OnClickedTrySpawnInvisible()
+        {
+            TrySpawnInvisible();
+        }
+
+        async void TrySpawnInvisible()
+        {
+            var randomPrefab = m_DynamicPrefabReferences[Random.Range(0, m_DynamicPrefabReferences.Count)];
+            await SpawnImmediatelyAndHideUntilPrefabIsLoadedOnClient(randomPrefab.AssetGUID, Random.insideUnitCircle * 5, Quaternion.identity);
+        }
+        
+        /// <summary>
+        /// This call spawns an addressable prefab by it's guid. It does not ensure that all the clients have loaded the
+        /// prefab before spawning it. All spawned objects are invisible to clients that don't have the prefab loaded.
+        /// The server tells the clients that lack the preloaded prefab to load it and acknowledge that they've loaded
+        /// it, and then the server makes the object visible to that client.
+        /// </summary>
+        /// <param name="guid"></param>
+        /// <returns></returns>
+        async Task<NetworkObject> SpawnImmediatelyAndHideUntilPrefabIsLoadedOnClient(string guid, Vector3 position, Quaternion rotation)
+        {
+            if (IsServer)
+            {
+                var assetGuid = new AddressableGUID()
+                {
+                    Value = guid
+                };
+                
+                return await Spawn(assetGuid);
+            }
+
+            return null;
+
+            async Task<NetworkObject> Spawn(AddressableGUID assetGuid)
+            {
+                var prefab = await DynamicPrefabLoadingUtilities.LoadDynamicPrefab(assetGuid, 
+                    m_ArtificialDelayMilliseconds);
+                var obj = Instantiate(prefab, position, rotation).GetComponent<NetworkObject>();
+                
+                if (m_PrefabHashToNetworkObjectId.TryGetValue(assetGuid.GetHashCode(), out var networkObjectIds))
+                {
+                    networkObjectIds.Add(obj);
+                }
+                else
+                {
+                    m_PrefabHashToNetworkObjectId.Add(assetGuid.GetHashCode(), new HashSet<NetworkObject>() {obj});
+                }
+
+                obj.CheckObjectVisibility = (clientId) => 
+                {
+                    //if the client has already loaded the prefab - we can make the object visible to them
+                    if (DynamicPrefabLoadingUtilities.HasClientLoadedPrefab(clientId, assetGuid.GetHashCode()))
+                    {
+                        return true;
+                    }
+                    //otherwise the clients need to load the prefab, and after they ack - the ShowHiddenObjectsToClient 
+                    LoadAddressableClientRpc(assetGuid, new ClientRpcParams(){Send = new ClientRpcSendParams(){TargetClientIds = new ulong[]{clientId}}});
+                    return false;
+                };
+                
+                obj.SpawnWithOwnership(m_NetworkManager.LocalClientId);
```

Comment:
why is ownership required here? I'd expect this Spawn method to only run server side

Reply:
Yeah the lines above gate this piece of code to only run on the server. This doesn't need to be here. I'll fix.